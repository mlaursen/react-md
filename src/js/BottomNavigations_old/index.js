import BottomNavigation from './BottomNavigation';
export default BottomNavigation;
