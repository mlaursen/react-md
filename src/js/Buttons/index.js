import Button from './Button';
export default Button;

export FlatButton from './FlatButton';
export RaisedButton from './RaisedButton';
export FloatingButton from './FloatingButton';
export IconButton from './IconButton';
