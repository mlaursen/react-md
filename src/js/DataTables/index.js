export DataTable from './DataTable';
export TableHeader from './TableHeader';
export TableBody from './TableBody';
export TableRow from './TableRow';
export TableColumn from './TableColumn';
export EditDialogColumn from './EditDialogColumn';
export TablePagination from './TablePagination';
export SelectFieldColumn from './SelectFieldColumn';
