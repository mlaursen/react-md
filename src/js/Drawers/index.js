import Drawer from './Drawer';
export default Drawer;
