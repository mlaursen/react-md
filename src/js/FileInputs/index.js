import FileInput from './FileInput';
export default FileInput;

import FileUpload from './FileUpload';

export { FileInput };
export { FileUpload };
