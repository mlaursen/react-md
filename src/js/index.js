import Autocomplete from './Autocompletes';
import Avatar from './Avatars';
import BottomNavigation from './BottomNavigations';
import Button, {
  FlatButton,
  FloatingButton,
  IconButton,
  RaisedButton,
} from './Buttons';
import {
  Card,
  CardActions,
  CardActionOverlay,
  CardHeader,
  CardMedia,
  CardText,
  CardTitle,
} from './Cards';
import Chip from './Chips';
import {
  DataTable,
  TableHeader,
  TableBody,
  TableRow,
  TableColumn,
  EditDialogColumn,
} from './DataTables';
import Dialog from './Dialogs';
import Divider from './Dividers';
import { SpeedDial } from './FABTransitions';
import FileInput, { FileUpload } from './FileInputs';
import FontIcon from './FontIcons';
import injectInk from './Inks';
import { List, ListItem } from './Lists';
import AccessibleFakeButton from './Helpers/AccessibleFakeButton';
import IconSeparator from './Helpers/IconSeparator';
import Menu from './Menus';
import NavigationDrawer from './NavigationDrawers';
import Paper from './Papers';
import { DatePicker, TimePicker } from './Pickers';
import { CircularProgress, LinearProgress } from './Progress';
import { Checkbox, Radio, RadioGroup, Switch } from './SelectionControls';
import SelectField from './SelectFields';
import Slider from './Sliders';
import Sidebar from './Sidebars';
import Snackbar from './Snackbars';
import Subheader from './Subheaders';
import { Tab, Tabs } from './Tabs';
import TextField from './TextFields';
import Toolbar from './Toolbars';
import injectTooltip from './Tooltips';

export { AccessibleFakeButton };
export { Autocomplete };
export { Avatar };
export { Button };
export { BottomNavigation };
export { Card };
export { CardActions };
export { CardActionOverlay };
export { CardHeader };
export { CardMedia };
export { CardText };
export { CardTitle };
export { Checkbox };
export { Chip };
export { CircularProgress };
export { DatePicker };
export { DataTable };
export { Dialog };
export { Divider };
export { EditDialogColumn };
export { FlatButton };
export { FloatingButton };
export { FileInput };
export { FileUpload };
export { FontIcon };
export { IconButton };
export { injectInk };
export { IconSeparator };
export { LinearProgress };
export { List };
export { ListItem };
export { Menu };
export { NavigationDrawer };
export { Paper };
export { Radio };
export { RadioGroup };
export { RaisedButton };
export { Slider };
export { Sidebar };
export { Snackbar };
export { SpeedDial };
export { Subheader };
export { Switch };
export { Tab };
export { TableBody };
export { TableColumn };
export { TableHeader };
export { TableRow };
export { Tabs };
export { TextField };
export { TimePicker };
export { Toolbar };
export { injectTooltip };

export default {
  Autocomplete,
  Avatar,
  BottomNavigation,
  Button,
  AccessibleFakeButton,
  FlatButton,
  RaisedButton,
  FloatingButton,
  IconButton,
  Card,
  CardHeader,
  CardTitle,
  CardMedia,
  CardActions,
  CardText,
  CardActionOverlay,
  DatePicker,
  DataTable,
  EditDialogColumn,
  List,
  ListItem,
  IconSeparator,
  CircularProgress,
  LinearProgress,
  Checkbox,
  RadioGroup,
  Radio,
  Switch,
  Tabs,
  Tab,
  Chip,
  Dialog,
  Divider,
  FileInput,
  FileUpload,
  FontIcon,
  Menu,
  NavigationDrawer,
  injectInk,
  Paper,
  SelectField,
  Slider,
  Sidebar,
  Snackbar,
  SpeedDial,
  Subheader,
  TableBody,
  TableColumn,
  TableHeader,
  TableRow,
  TextField,
  TimePicker,
  Toolbar,
  injectTooltip,
};
