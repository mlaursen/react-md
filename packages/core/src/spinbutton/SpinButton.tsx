"use client";

import { type HTMLAttributes, forwardRef } from "react";

import { type LabelRequiredForA11y } from "../types.js";
import { defaultGetSpinButtonTextContent } from "./defaults.js";
import {
  type SpinButtonOptions,
  type SpinButtonUncontrolledValueOptions,
  type SpinButtonValueOptions,
} from "./types.js";
import { useSpinButton } from "./useSpinButton.js";

/**
 * @since 6.4.0
 */
export interface SpinButtonProps
  extends Omit<HTMLAttributes<HTMLDivElement>, keyof SpinButtonOptions>,
    Omit<SpinButtonOptions<HTMLDivElement>, "ref" | "children"> {}

/**
 * The `SpinButton` component can be used as a simple wrapper around the
 * {@link useSpinButton} hook if it should be rendered within a `<div>` and
 * only the rendering behavior is required.
 *
 * @since 6.4.0
 */
export const SpinButton = forwardRef<
  HTMLDivElement,
  LabelRequiredForA11y<SpinButtonProps> & SpinButtonValueOptions
>(function SpinButton(props, ref) {
  const {
    id,
    min,
    max,
    minDigits,
    maxDigits,
    step,
    form,
    error,
    disabled,
    readOnly,
    required,
    fallback,
    mappings,
    value,
    defaultValue,
    onValueChange,
    getValueText,
    getTextContent = defaultGetSpinButtonTextContent,
    placeholderChar,
    defaultKeyboardValue,
    ...remaining
  } = props;

  const valueProps = {
    value,
    defaultValue,
  } as SpinButtonUncontrolledValueOptions<HTMLDivElement>;

  const { value: currentValue, spinButtonProps } = useSpinButton({
    id,
    ref,
    min,
    max,
    minDigits,
    maxDigits,
    step,
    form,
    fallback,
    mappings,
    error,
    disabled,
    readOnly,
    required,
    ...valueProps,
    onValueChange,
    getValueText,
    getTextContent,
    placeholderChar,
    defaultKeyboardValue,
  });

  return (
    <div {...remaining} {...spinButtonProps}>
      {getTextContent({
        min,
        max,
        minDigits,
        maxDigits,
        value: currentValue,
        fallback,
        placeholderChar,
      })}
    </div>
  );
});
