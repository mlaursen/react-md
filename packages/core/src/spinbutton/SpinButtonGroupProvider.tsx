"use client";

import { createContext, useContext } from "react";

import { type KeyboardMovementContext } from "../movement/types.js";
import { DEFAULT_KEYBOARD_MOVEMENT_CONTEXT } from "../movement/useKeyboardMovementProvider.js";

/**
 * @since 6.4.0
 */
export type SpinButtonGroupContext = KeyboardMovementContext;

const context = createContext<SpinButtonGroupContext>(
  DEFAULT_KEYBOARD_MOVEMENT_CONTEXT
);
context.displayName = "SpinButtonGroup";

/**
 * This should be used along with the `useSpinButtonGroupProvider` to link
 * `SpinButton` components together. The next `SpinButton` will be focused as
 * an value is typed to completion.
 *
 * @since 6.4.0
 */
export const { Provider: SpinButtonGroupProvider } = context;

/**
 * @since 6.4.0
 */
export function useSpinButtonGroup(): Readonly<SpinButtonGroupContext> {
  return useContext(context);
}
