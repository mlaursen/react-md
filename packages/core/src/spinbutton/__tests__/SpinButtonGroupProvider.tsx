import { describe, expect, it, jest } from "@jest/globals";
import { type ReactElement, createRef } from "react";

import { Box } from "../../box/Box.js";
import { render, screen, userEvent } from "../../test-utils/index.js";
import { SpinButton } from "../SpinButton.js";
import { SpinButtonGroupProvider } from "../SpinButtonGroupProvider.js";
import {
  type SpinButtonGroupProviderOptions,
  useSpinButtonGroupProvider,
} from "../useSpinButtonGroupProvider.js";

interface TestProps {
  options?: SpinButtonGroupProviderOptions;
}

function Test({ options }: TestProps): ReactElement {
  const { movementProps, movementContext } =
    useSpinButtonGroupProvider(options);

  return (
    <Box {...movementProps}>
      <SpinButtonGroupProvider value={movementContext}>
        <SpinButton aria-label="Hours" min={1} max={12} fallback="HH" />
        <SpinButton aria-label="Minutes" min={0} max={60} fallback="MM" />
      </SpinButtonGroupProvider>
    </Box>
  );
}

const setup = (options?: SpinButtonGroupProviderOptions) => {
  const user = userEvent.setup();
  render(<Test options={options} />);

  return {
    user,
    group: screen.getByRole("group"),
    hours: screen.getByRole("spinbutton", { name: "Hours" }),
    minutes: screen.getByRole("spinbutton", { name: "Minutes" }),
  };
};

describe("SpinButtonGroupProvider", () => {
  it("should be able to connect a group of SpinButtons together and focus the next when typed-to-completion", async () => {
    const { user, hours, minutes } = setup();

    await user.tab();
    expect(hours).toHaveFocus();

    await user.keyboard("3");
    expect(hours).toHaveValue(3);
    expect(minutes).toHaveFocus();

    await user.keyboard("00");
    expect(minutes).toHaveValue(0);
    expect(minutes).toHaveFocus();
  });

  it("should allow for the event listeners to be hooked into", async () => {
    const ref = createRef<HTMLDivElement>();
    const onClick = jest.fn();
    const onFocus = jest.fn();
    const onKeyDown = jest.fn();
    const { user, group, hours } = setup({
      ref,
      onClick,
      onFocus,
      onKeyDown,
    });

    expect(ref.current).toBeInstanceOf(HTMLDivElement);
    expect(ref.current).toBe(group);
    expect(group).toMatchSnapshot();
    expect(group).not.toHaveAttribute("tabIndex");
    expect(group).not.toHaveAttribute("aria-activedescendant");

    await user.type(hours, "1");

    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onFocus).toHaveBeenCalledTimes(1);
    expect(onKeyDown).toHaveBeenCalledTimes(1);
  });
});
