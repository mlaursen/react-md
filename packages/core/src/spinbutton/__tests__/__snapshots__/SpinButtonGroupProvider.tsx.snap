// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinButtonGroupProvider should allow for the event listeners to be hooked into 1`] = `
<div
  class="rmd-box rmd-box--gap rmd-box--wrap rmd-box--padded"
  role="group"
>
  <div
    aria-label="Hours"
    aria-valuemax="12"
    aria-valuemin="1"
    aria-valuetext="No value selected"
    autocapitalize="none"
    autocorrect="off"
    contenteditable="true"
    id="spinbutton-:r2:"
    inputmode="numeric"
    role="spinbutton"
    spellcheck="false"
    tabindex="0"
  >
    HH
  </div>
  <div
    aria-label="Minutes"
    aria-valuemax="60"
    aria-valuemin="0"
    aria-valuetext="No value selected"
    autocapitalize="none"
    autocorrect="off"
    contenteditable="true"
    id="spinbutton-:r3:"
    inputmode="numeric"
    role="spinbutton"
    spellcheck="false"
    tabindex="0"
  >
    MM
  </div>
</div>
`;
